lib/read.js

```diff
@@ -91,9 +97,9 @@ function read (req, res, next, parse, debug, options) {
 
       // read off entire request
       stream.resume()
-      onFinished(req, function onfinished () {
+      onFinished(req, preserveAsyncContext('OnFinished', function onfinished () {
```

Comment:
Should not the `onFinished` function itself be preserving the async context? Wouldn't adding it within `onFinished` function directly preserve the context for every dependency on npm that uses it instead of needing to work through every one of them (like this module) to add this? It seems quite burdensome for all the dependencies of that module if this is is true. Because by that logic, why don't you do this around the call to `bodyParser` itself?

Reply:
That's true, in principle this is the right thing to do, and also that's the reason why the discussion on how to handle async context in event handlers is happening now between NodeJS folks, as it would be the lowest layer that benefits all the dependent modules.

In practice though, having contributed to third party modules in the past, I found it quite difficult to get the attention of the maintainers to even look at the PR, let alone merging it. So I figured `body-parser` to be my best shot, as it was a single direct dependency of my apps that was causing issues (rather than two transitive ones)  and it looked alive on GitHub.

Having said that, I see that you're a collaborator on both of these, so I presume that's not the problem here :slightly_smiling_face: If you're willing to pull in my PRs there, release new versions, let me update their declarations in `body-parser` and then release it, I'm happy to do it and we can close this PR.

Reply:
Yep, same as getBody, I'd recommend this be added to onFinished.